File: archive/go/practices/linkedlist/rotate-list.go
package linkedlist

import . "leetcode/common"

func rotateRight(head *ListNode, k int) *ListNode {
	run := head
	var l = 0
	for run != nil {
		l++
		run = run.Next
	}
	k = k / l
	run = head
	var p ListNode
	for i := 0; i < l-k; i++ {
		run = run.Next
	}
	p = *run
	for run.Next != nil {
		var tmp = run.Next
		run.Next = &p
		p = *run
		run = tmp
	}
	return &p
}

func _rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k == 0 {
		return head
	}
	var l = 1
	var newH, tail *ListNode
	newH, tail = head, head
	for tail.Next != nil {
		tail = tail.Next
		l++
	}
	// make linked list circular
	tail.Next = head
	if (k % l) != 0 {
		// find the rotation point
		for i := 0; i < l-k; i++ {
			tail = tail.Next
		}
	}
	newH = tail.Next
	tail.Next = nil
	return newH
}

File: archive/java/src/main/java/cn/az/code/year2020/oct/RotateList.java
package cn.az.code.year2020.oct;

import cn.az.code.common.ListNode;

/**
 * @author az
 * @since 10/09/20
 */
public class RotateList {

    public ListNode rotate(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy, slow = dummy;

        int i;
        for (i = 0; fast.next != null; i++) {
            fast = fast.next;
        }

        for (int j = i - k % i; j > 0; j--) {
            slow = slow.next;
        }

        fast.next = dummy.next; //Do the rotation
        dummy.next = slow.next;
        slow.next = null;

        return dummy.next;
    }

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        int len = 1;
        ListNode newHead, tail = head;
        while (tail.next != null) {
            tail = tail.next;
            len++;
        }
        k %= len;
        tail.next = head;
        for (int i = 0; i < len - k; i++) {
            tail = tail.next;
        }
        newHead = tail.next;
        tail.next = null;
        return newHead;
    }
}
